### PR TITLE
Field documentation

### PR DESCRIPTION
**Pull request by edburyenegren-okta**
This adds docs and stories for Field. It also updates related stories so they can be utilized as examples for Field.

### CODE REVIEW

**Comment by edburyenegren-okta** [comment]
Note, I'm seeing differences locally and on CI:

Screen Shot 2021-12-09 at 1 42 05 PM

Screen Shot 2021-12-09 at 1 42 13 PM

I don't believe this delta is related to anything specific to Field, and I'd prefer to handle this discrepancy in another ticket.

File: packages/odyssey-storybook/src/components/Field/Field.mdx [hunk 1]
```diff
@@ -1,43 +1,150 @@
...
 # Field
 
-These captions help make forms more accessible by providing context to the user. They can be used with all Odyssey inputs.
+Field provides all of the supporting context required by inputs, including labels, hints, and errors.
+
+This component is also responsible for managing the appropriate application of `aria-` attributes and `id` associations between Inputs and their associated elements.
+
+[View source on GitHub](https://github.com/okta/odyssey/tree/master/packages/odyssey-react/src/components/Field)
 
 ## Anatomy
```
Comment (jeffbelser-okta) [inline packages/odyssey-storybook/src/components/Field/Field.mdx:11]: can we include the prop table up top here for engineers looking for the API? It appears to be missing here in the docs page entirely.

Incidentally, you can also fix the broken prop tables for Field in this PR which would be great. [We're using a new approach described here.](https://okta.slack.com/archives/G01JS0FE32P/p1638891117039900?thread_ts=1638890977.039800&cid=G01JS0FE32P)

File: packages/odyssey-storybook/src/components/TextInput/TextInput.stories.tsx [hunk 2]
```diff
@@ -50,8 +50,20 @@ export default {
...
+Text.args = {};
+
+export const Optional = Template.bind({});
+Optional.args = {
+  required: false,
+  optionalLabel: "Optional",
+};
+
+export const KitchenSink = Template.bind({});
+KitchenSink.args = {
```
Comment (jeffbelser-okta) [inline packages/odyssey-storybook/src/components/TextInput/TextInput.stories.tsx:62]: consider a story title that isn't as colloquial, something with meaning to a broader audience. Perhaps Survey or your use of Anatomy in the written docs above?

File: packages/odyssey-storybook/src/components/Field/Field.stories.tsx [hunk 3]
```diff
@@ -46,5 +49,32 @@ export default {
 
 const Template: Story<FieldProps> = (props) => <Field {...props} />;
 
-export const FieldSolo = Template.bind({});
-FieldSolo.args = {};
+export const KitchenSink = Template.bind({});
+KitchenSink.args = {
```
Comment (jeffbelser-okta) [inline packages/odyssey-storybook/src/components/Field/Field.stories.tsx:53]: see note on title below

File: packages/odyssey-storybook/src/components/Field/Field.mdx [hunk 4]
```diff
@@ -1,43 +1,150 @@
...
+
+The label also indicates the Field's optional/required status to ensure screen readers properly announce the state.
+
+<Canvas>
+  <Story id="components-field--field-optional" />
+</Canvas>
+
+### Hint
+
+A hint is helper text for the Field. The field hint is similar to the disallowed `placeholder` attribute. It should provide any necessary help text for the Field but is optional.
```
Comment (jeffbelser-okta) [inline packages/odyssey-storybook/src/components/Field/Field.mdx:33]: disallowed placeholder attribute? if we don't want placeholders (which I wasn't aware of!) we should prevent their use in code too. Can you update the types for our input components to make this so before we merge the docs in this form? Alternatively, leave this line of docs off and we'll update everything in another subsequent PR at a later point.

File: packages/odyssey-storybook/src/components/Field/Field.stories.tsx [hunk 5]
```diff
@@ -28,11 +28,14 @@ export default {
   args: {
     label: "Destination",
     optionalLabel: "Optional",
```
Comment (jeffbelser-okta) [inline packages/odyssey-storybook/src/components/Field/Field.stories.tsx:30]: Having this as part of the default arguments here makes the code examples confusing for the Field docs, e.g.
```
<WithStyles(Field)
  hint="Your planetary destination."
  optionalLabel="Optional"
  required
/>
```

Consider removing this and including it only where it is applicable for all the stories below. We might need to update types as well to have the logic we want enforced in code, that is `optionalLabel` is only required to be provided when a Field is `required=false | undefined`.

File: packages/odyssey-storybook/src/components/Field/Field.stories.tsx [hunk 6]
```diff
@@ -28,11 +28,14 @@ export default {
...
+    hint: "Your planetary destination.",
     error: (
       <>
-        <ScreenReaderText>Error:</ScreenReaderText> Descriptive error text.
+        <ScreenReaderText>Error:</ScreenReaderText> This field may not be left
+        blank.
       </>
     ),
+    required: true,
```
Comment (jeffbelser-okta) [inline packages/odyssey-storybook/src/components/Field/Field.stories.tsx:38]: same here for `required` and `hint` - these feel like props that would be better suited to be included in each example below where applicable. Rather than overriding them with `undefined` again below...

File: packages/odyssey-storybook/src/components/Field/Field.mdx [hunk 7]
```diff
@@ -1,43 +1,150 @@
...
+
+## Content Guidelines
+
+### Labels
+
+A well-written label describes the purpose of the Field for the user.
+
+Use concise, descriptive labels (a word or two) to help users quickly scan your form.
+
+Finally, please use sentence casing for your labels. Sentence case is easier to read and will help reduce error rates and speed up form completion time.
```
Comment (jeffbelser-okta) [inline packages/odyssey-storybook/src/components/Field/Field.mdx:111]: an example will help here

File: packages/odyssey-storybook/src/components/Field/Field.mdx [hunk 8]
```diff
@@ -1,43 +1,150 @@
...
+
+By default, inputs and Fields are set to `required`. As the default, this state has no unique affordance.
+
+<Canvas>
+  <Story id="components-textinput--text" />
+</Canvas>
+
+### Optional
+
+Fields with `required` set to `false` will be marked as optional:
```
Comment (jeffbelser-okta) [inline packages/odyssey-storybook/src/components/Field/Field.mdx:89]: consider rewording to: explicitly set to false. It is the only way to mark a field optional

File: packages/odyssey-storybook/src/components/Field/Field.mdx [hunk 9]
```diff
@@ -1,43 +1,150 @@
...
+Remember: a good label may eliminate the need for a hint.
 
-Keep labels to a word or two so users can quickly scan the form. Always use sentence casing.
+### Errors
 
-**STORY MISSING**
+Error messaging should explain: (1) what has gone wrong and (2) how the user can fix the issue.
 
-### Field hint
+"This value is invalid." is less helpful than "This field requires a complete email address."
```
Comment (jeffbelser-okta) [inline packages/odyssey-storybook/src/components/Field/Field.mdx:129]: the more helpful example here contradicts the guidelines above, it doesn't explain what has gone wrong. Should we reword this to instead say something like this?

Error messaging should at a minimum explain what has gone wrong and at best, explain how the user can fix the issue.

File: packages/odyssey-storybook/src/components/Field/Field.mdx [hunk 10]
```diff
@@ -1,43 +1,150 @@
...
+Your error messages should include an "Error:" label, whether visible or visually hidden.
 
-Invalid fields should include an inline message to explain the error.
+## Globalization
 
-If possible, describe how the error may be resolved.
+To support ease of localization, Field's `optionalLabel` string has no default and must be supplied by the user.
 
-**STORY MISSING**
+Similarly, an "Error:" label is not automatically inserted into Field error messages.
```
Comment (jeffbelser-okta) [inline packages/odyssey-storybook/src/components/Field/Field.mdx:143]: the globalization notes here feel superfluous - I'm not a fan of documenting what our components do not do. Perhaps reword this to simply mention that all user facings strings, be it labels, hints, placeholders, or error text are the responsibility of the consuming developer to translate? This may already be apparent for some developers.

**Review by jeffbelser-okta — COMMENTED** [review]
strong work 💪 some thoughts inline

File: packages/odyssey-storybook/src/components/Field/Field.mdx (hunk 1 above)
Comment (edburyenegren-okta) [reply packages/odyssey-storybook/src/components/Field/Field.mdx:11]: The prop table is just below, but wasn't populating before. Added a `Field.docgen` - working great now!

**Review by arnoldsandoval-okta — APPROVED** [review]

File: packages/odyssey-storybook/src/components/Field/Field.stories.tsx (hunk 5 above)
Comment (edburyenegren-okta) [reply packages/odyssey-storybook/src/components/Field/Field.stories.tsx]: I've reworked the stories here to not utilize default args since `required` is the only default we expect (per the component).

Agreed on the internal logic around `optionalLabel` - should be required when the Field/Input isn't. Ticket here, so this doesn't get lost: https://oktainc.atlassian.net/browse/OKTA-453308

File: packages/odyssey-storybook/src/components/Field/Field.mdx (hunk 4 above)
Comment (edburyenegren-okta) [reply packages/odyssey-storybook/src/components/Field/Field.mdx:33]: "disallowed" is definitely too strong here, as there are places that `placeholder` is an existing escape hatch. I'll update the language. 

As guidance, we want designers and developers to avoid `placeholder` as it comes with a11y and usability issues.

File: packages/odyssey-storybook/src/components/Field/Field.mdx (hunk 7 above)
Comment (edburyenegren-okta) [reply packages/odyssey-storybook/src/components/Field/Field.mdx:113]: Examples added!

General note: I'm sticking to MDX for content examples until we have a solution for non-CSF stories or hiding "content" stories from the sidebar/VRT. Ideally, these will be in-component examples soon.

File: packages/odyssey-storybook/src/components/Field/Field.docgen.tsx [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import React, { forwardRef } from "react";
+import type { FieldProps } from "@okta/odyssey-react";
+
+const Source = forwardRef<HTMLDivElement, FieldProps>(() => <></>);
```
Comment (jeffbelser-okta) [inline packages/odyssey-storybook/src/components/Field/Field.docgen.tsx:16]: lets rename this to something more meaningful like `FieldDocGen` and export it below on line 18 - I wanted to make the same change for `Box.

Also, no need for the `forwardRef` here, all we need is a function component that accepts the same props as `Field` and returns `null`, we don't even need the react fragment

Comment (arnoldsandoval) [reply packages/odyssey-storybook/src/components/Field/Field.docgen.tsx:16]: Why do these docgen files exist at all? They feel completely unnecessary on the surface, does this have to do with complexity in our TypeScript implementation?

Comment (jeffbelser-okta) [reply packages/odyssey-storybook/src/components/Field/Field.docgen.tsx:16]: @arnoldsandoval the [tool](https://github.com/styleguidist/react-docgen-typescript) we're using to create prop tables in our docs doesn't understand some of the component composition patterns we're using, forwardRefWithStatics being one, polymorphic props being another. Rather than trying to continue to tinker with our component source we're going to setup docgen only components within `@okta/odyssey-storybook` to make parsing out prop tables straightforward for the parser. Open to other suggestions or ideas too

File: packages/odyssey-storybook/src/components/Field/Field.mdx (hunk 10 above)
Comment (arnoldsandoval) [reply packages/odyssey-storybook/src/components/Field/Field.mdx:147]: Designers, PMs and others may not know what the components do. I don't think being explicit about behavior is a bad thing.
